### PR TITLE
Update Lando to 3.0.0-rc.2+.

### PR DESCRIPTION
**Pull request by tormi**
## Features

- Updates Lando to 3.0.0-rc.2+ (breaking changes, see **Attention!**)
- Introduces [Lando base file](https://docs.devwithlando.io/config/lando.html#base-file) concept.
- Adds out of the box support for following testing tools (see readme for details):
  - Codeception
  - PHP_CodeSniffer
  - PHPStan
  - PHPUnit
- Adds support for automatic coding standards detection using [Dealerdirect/phpcodesniffer-composer-installer](https://github.com/DealerDirect/phpcodesniffer-composer-installer) package.
- Adds Yarn and npm support to Lando tooling.
- Adds `composer install` and `yarn` to `lando start` and `lando rebuild` commands.

### Attention!
- This requires Lando 3.0.0-rc.2+, read about the changes here: https://thinktandem.io/blog/2019/02/01/lando-is-ready-for-the-masses-with-rc2-release/.
- Perform `lando db-import` before the update to backup your local database.
- Perform `lando destroy` before `lando start` to destroy old abandoned images, read more here: https://docs.devwithlando.io/guides/updating-to-rc2.html#new-images
- Delete `~/.lando` folder before `lando start` just in case, read about the background here: https://docs.devwithlando.io/guides/updating-to-rc2.html#removing-lando.

### CODE REVIEW

**Comment by tormi** [comment]
@mglaman's article about testing Drupal with PHPStan: https://glamanate.com/blog/testing-your-drupal-code-base-deprecated-code-usage-phpstan

**Comment by tormi** [comment]
Good introduction to PHP Testing with Codeception: https://www.toptal.com/php/php-testing-with-codeception

File: .lando.base.yml [hunk 1]
```diff
@@ -40,6 +40,10 @@ tooling:
     service: appserver
     description: Runs composer update drupal/core webflo/drupal-core-require-dev --with-dependencies
     cmd: composer update drupal/core webflo/drupal-core-require-dev --with-dependencies
+  dsi:
+    service: appserver
+    description: Runs drush site install with drupal8 recipe's --db-url
+    cmd: /app/vendor/bin/drush si --db-url=mysql://drupal8:drupal8@database:3306/drupal8
```
Comment (floretan) [inline .lando.base.yml]: Do we really need to provide the database credentials? These should already be provided via the lando.settings.php.

In most cases people where people need to install drupal, they will either use config_installer, or if they are starting from scratch they should make a conscious choice of installation profile.

File: .lando.base.yml [hunk 2]
```diff
@@ -0,0 +1,78 @@
+name: Silta
+recipe: drupal8
+
+config:
+  webroot: web
+  via: nginx
+  php: '7.1'
```
Comment (floretan) [inline .lando.base.yml]: We should probably use 7.2 as the default for new projects.

File: composer.json [hunk 3]
```diff
@@ -32,7 +32,13 @@
         "wunderio/drupal-ping": "^1.0"
     },
     "require-dev": {
-        "webflo/drupal-core-require-dev": "~8.5"
+        "codeception/codeception": "^2.5",
```
Comment (floretan) [inline composer.json]: Having all these dependencies in the project rather than the CI environments makes a lot of sense.

File: README.md [hunk 4]
```diff
@@ -21,3 +20,45 @@ Have a look at the file for details, but in short this is how it works:
...
 - Install or update our helm chart while passing our custom images as parameters.
 - The helm chart executes the usual drush deployment commands.
+
+## Local development
+
+Install [Lando](https://docs.devwithlando.io/). Add your project's name and other project-specific settings to `.lando.yml` and run `lando start`.
+
+Useful helper commands:
+
+- `lando cre` runs `composer require`
```
Comment (floretan) [inline README.md]: I'm personally not a fan of wrapping common commands with an alias just to save keystrokes. It's confusing for newcomers, and it hides what's happening.

File: README.md [hunk 5]
```diff
@@ -21,3 +20,45 @@ Have a look at the file for details, but in short this is how it works:
...
+- `lando cupdc` runs `composer update drupal/core webflo/drupal-core-require-dev --with-dependencies`
+- `lando dsi` runs `drush si --db-url=mysql://drupal8:drupal8@database:3306/drupal8`. Run `lando dsi demo_umami -y` to install `demo_umami` profile.
+
+Run `lando` for complete list of available Lando commands.
+
+### Testing
+
+Supported testing tools:
+
+#### [PHP_CodeSniffer](https://github.com/squizlabs/PHP_CodeSniffer) with Drupal, DrupalPractice standards
```
Comment (floretan) [inline README.md]: We actually have our own phpcs.xml configuration, which extends the Drupal standard to make it a bit more relaxed where it makes sense.

File: README.md [hunk 6]
```diff
@@ -21,3 +20,45 @@ Have a look at the file for details, but in short this is how it works:
...
+Examples:
+
+- `lando phpcs web/modules/custom`
+- `lando phpcbf web/modules/custom`
+
+#### [Codeception](https://github.com/Codeception/Codeception)
+
+Usage: `lando codeception <command>`
+
+Create Codeception tests folder by running `lando codeception bootstrap`.
```
Comment (floretan) [inline README.md]: We should use the same default setup that Guntis has created for the WunderTools repository.

File: phpstan.neon [hunk 7]
```diff
@@ -0,0 +1,15 @@
+parameters:
+	customRulesetUsed: true
+	reportUnmatchedIgnoredErrors: false
+	# Ignore phpstan-drupal extension's rules.
+	ignoreErrors:
+		- '#\Drupal calls should be avoided in classes, use dependency injection instead#'
```
Comment (floretan) [inline phpstan.neon]: I don't understand this notation. Does that mean that these rules are ignored with the current configuration, or are they commented out?

File: .lando.base.yml [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    type: mailhog
+    hogfrom:
+      - appserver
+
+  appserver:
+    overrides:
+        environment:
+          HASH_SALT: notsosecurehash
+
+  # chrome:
```
Comment (floretan) [inline .lando.base.yml]: Having this service in the same file makes a lot of sense.

File: phpstan.neon (hunk 7 above)
Comment (mglaman) [reply phpstan.neon]: It's regex to ignore error output.

File: .lando.base.yml (hunk 8 above)
Comment (tormi) [reply .lando.base.yml]: You mean Chrome? We should refactor that as well.

File: README.md (hunk 4 above)
Comment (tormi) [reply README.md]: Ok, I'll remove those ;)

File: phpstan.neon (hunk 7 above)
Comment (tormi) [reply phpstan.neon]: Thanks, Matt for your input and phpstan-drupal extension!

File: .lando.base.yml (hunk 1 above)
Comment (tormi) [reply .lando.base.yml]: Removed wrapper commands.

File: .lando.base.yml (hunk 8 above)
Comment (floretan) [reply .lando.base.yml]: A refactor could be useful, but that's for another time.

Comment (tormi) [reply .lando.base.yml]: > Having this service in the same file makes a lot of sense.
I see, you were talking about `lando-docker-compose.yml` > `.lando.base.yml` :)

File: README.md (hunk 6 above)
Comment (tormi) [reply README.md]: `lando codeception run acceptance` and `lando codeception run unit` seem to be working now, `lando codeception run functional` gives an error:

```
Codeception PHP Testing Framework v2.5.4
Powered by PHPUnit 6.5.14 by Sebastian Bergmann and contributors.
Running with seed:

In Drupal.php line 130:

  \Drupal::$container is not initialized yet. \Drupal::setContainer() must be called with a real container.
```

Unfortunately, I'm just in the beginning of becoming a testing guru, so some help is needed here - @guncha25, @floretan :)

Comment (tormi) [reply README.md]: Thanks, @smmccabe for the 💯`crap` 💯  here: https://github.com/lando/lando/issues/1090#issuecomment-419269001 :)

File: README.md (hunk 5 above)
Comment (tormi) [reply README.md]: Any ideas of how to automatically set `installed_paths` for this config? I introduced https://github.com/DealerDirect/phpcodesniffer-composer-installer earlier and it doesn't pick up..

File: README.md (hunk 6 above)
Comment (tormi) [reply README.md]: This is probably ok for now and needs project-specific adjustments.

**Comment by floretan** [comment]
There are still some phpcs warnings from the test files, but otherwise this looks good!

**Comment by tormi** [comment]
Should we fix those?

**Comment by tormi** [comment]
This seems to be something we can't `fix`, can we? Part of the `phpcs tests` output:

```
FILE: /app/tests/functional/ExampleFunctionalCest.php
------------------------------------------------------------------------------------------------------------
FOUND 6 ERRORS AFFECTING 6 LINES
------------------------------------------------------------------------------------------------------------
  3 | ERROR | Doc comment is empty
  8 | ERROR | Missing short description in doc comment
  9 | ERROR | Missing parameter comment
 11 | ERROR | Variable "I" starts with a capital letter, but only $lowerCamelCase or $snake_case is allowed
 12 | ERROR | Variable "I" starts with a capital letter, but only $lowerCamelCase or $snake_case is allowed
 13 | ERROR | Variable "I" starts with a capital letter, but only $lowerCamelCase or $snake_case is allowed
------------------------------------------------------------------------------------------------------------
```

**Comment by tormi** [comment]
We have following standards available thanks to dealerdirect/phpcodesniffer-composer-installer: 
`lando phpcs -i`
_The installed coding standards are PEAR, Zend, PSR2, MySource, Squiz, PSR1, PSR12, Drupal, DrupalPractice and codeception_

I should probably use `codeception` standard, instead ;) This requires removing `--standard=Drupal,DrupalPractice` from `lando phpcs` and `lando phpcbf` tools.

**Comment by tormi** [comment]
@floretan , we should use `codeception` standard in CircleCI tests as well.

```
lando phpcbf --standard=codeception tests

PHPCBF RESULT SUMMARY
----------------------------------------------------------------------
FILE                                                  FIXED  REMAINING
----------------------------------------------------------------------
/app/tests/_support/Helper/Functional.php             1      0
/app/tests/_support/Helper/Unit.php                   1      0
/app/tests/_support/Helper/Acceptance.php             1      0
/app/tests/acceptance/ExampleAcceptanceCest.php       7      0
/app/tests/functional/ExampleFunctionalCest.php       7      0
----------------------------------------------------------------------
A TOTAL OF 17 ERRORS WERE FIXED IN 5 FILES
----------------------------------------------------------------------
```

**Comment by tormi** [comment]
@floretan , I think I'm done here.

File: README.md (hunk 5 above)
Comment (tormi) [reply README.md]: Resolved with 0aae671.

**Comment by tormi** [comment]
@guncha25 's great video about Drupal and Codeception: https://www.youtube.com/watch?v=A3yx88Lmz0E

**Comment by tormi** [comment]
Validation doesn't pass. I think we should remove hardcoded `phpcs --standard=phpcs.xml -s` from `orb.yml` and use `.circleci/config.yml` only for defining CircleCI testing standards because they are project-based.

**Comment by tormi** [comment]
2b2d391 was triggered too quickly by @forgenator work at https://github.com/wunderio/nivos/commit/8ce4564144c00d52d857527656742fa9a25651bb. Took a deep breath after push and reverted ;) But this was helpful because I found out we had to clean up even further: #30.

**Comment by tormi** [comment]
Or not: #31.

**Comment by Jancis** [comment]
@tormi, when this gets into drupal-project, please make another PR for https://github.com/wunderio/WunderTools repository. This is really valuable and we want this in wundertools too.

**Comment by tormi** [comment]
Sure, but I need some internal time budget for this then.

**Comment by tormi** [comment]
Somehow codeception standards are available locally, but not available in CI:

## Local Lando
```
Tormis-MacBook-Pro:drupal-project tormitabor$ lando phpcs -i
The installed coding standards are PEAR, Zend, PSR2, MySource, Squiz, PSR1, PSR12, Drupal, DrupalPractice and codeception
```

## CircleCI
```
#!/bin/bash -eo pipefail
phpcs --standard=codeception tests
ERROR: the "codeception" coding standard is not installed. The installed coding standards are Zend, PSR12, MySource, PSR2, Squiz, PSR1, PEAR, DrupalPractice and Drupal
Exited with code 3
```

**Comment by tormi** [comment]
Still not available..

```
#!/bin/bash -eo pipefail
phpcs --standard=codeception tests
ERROR: the "codeception" coding standard is not installed. The installed coding standards are MySource, PEAR, PSR1, PSR12, Zend, PSR2, Squiz, DrupalPractice and Drupal
Exited with code 3
```

**Comment by tormi** [comment]
`silta/drupal-composer-install` runs `composer install -n --prefer-dist --ignore-platform-reqs --no-dev --optimize-autoloader`.

Seems like we need `composer install` without `--no-dev` flag to install `dev` libraries which brings on additional standards (like codeception via `guncha25/drupal-codeception`):

```
    "require-dev": {
        "dealerdirect/phpcodesniffer-composer-installer": "*",
        "drupal/coder": "^8.3",
        "guncha25/drupal-codeception": "^8.0",
        "mglaman/phpstan-drupal": "^0.11.2",
        "phpstan/phpstan": "^0.11.4",
        "phpstan/phpstan-deprecation-rules": "^0.11.0",
        "webflo/drupal-core-require-dev": "^8.6.0"
    },
```

**Comment by tormi** [comment]
@floretan , we need to install additional PHP extensions to successfully run additional tests on CI: https://circleci.com/gh/wunderio/drupal-project/4097

```
Updating dependencies (including require-dev)
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - drupal/core 8.8.x-dev requires ext-gd * -> the requested PHP extension gd is missing from your system.
    - drupal/core 8.7.x-dev requires ext-gd * -> the requested PHP extension gd is missing from your system.
    - drupal/core 8.7.0-alpha2 requires ext-gd * -> the requested PHP extension gd is missing from your system.
    - drupal/core 8.7.0-alpha1 requires ext-gd * -> the requested PHP extension gd is missing from your system.
    - drupal/core 8.6.x-dev requires ext-gd * -> the requested PHP extension gd is missing from your system.
```

**Comment by tormi** [comment]
https://github.com/reload/ding2/commit/b06fbd25ee78e35ef5158eb2cef8e5f698da72e9

**Comment by tormi** [comment]
There are too much changes in this PR which needs some additional communication. I'll take the bare Lando upgrade part from this and push as a separate PR.

**Comment by tormi** [comment]
https://circleci.com/gh/wunderio/drupal-project/4127

`composer install -n --prefer-dist --ignore-platform-reqs --optimize-autoloader` step:

`PHP CodeSniffer Config installed_paths set to ../../codeception,../../drupal/coder/coder_sniffer` indicates that `codeception` standard is available.

`phpcs --standard=codeception tests`:

`ERROR: the "codeception" coding standard is not installed. The installed coding standards are MySource, Squiz, PEAR, PSR1, PSR12, Zend, PSR2, DrupalPractice and Drupal`

Screenshot 2019-03-25 at 22 37 02

**Comment by tormi** [comment]
Closing in favour of https://github.com/wunderio/drupal-project/pull/38. Enhancements in this PR shoud go in as a separate PR-s.

**Pull request closed without merging**